storage/src/commit_log/mod.rs

```diff
@@ -0,0 +1,127 @@
+use std::convert::TryInto;
+use std::path::{Path, PathBuf};
+use std::sync::{Arc, RwLock};
+
+use crate::commit_log::error::TezedgeCommitLogError;
+use crate::commit_log::reader::Reader;
+use crate::commit_log::writer::Writer;
+
+pub type CommitLogRef = Arc<RwLock<CommitLog>>;
+
+pub mod error;
+mod reader;
+mod writer;
+
+const INDEX_FILE_NAME: &str = "table.index";
+const DATA_FILE_NAME: &str = "table.data";
+
+const TH_LENGTH: usize = 16;
+
+pub type Message = Vec<u8>;
+
+pub struct MessageSet {
+    cursor: usize,
+    indexes: Vec<Index>,
+    acc: usize,
+    buf: Vec<u8>,
+}
+
+impl MessageSet {
+    fn new(indexes: Vec<Index>, buf: Vec<u8>) -> Self {
+        Self {
+            cursor: 0,
+            indexes,
+            acc: 0,
+            buf,
+        }
+    }
+}
+
+impl Iterator for MessageSet {
+    type Item = Message;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let index = self.indexes.get(self.cursor)?;
+        let data = &self.buf[self.acc..(self.acc + index.data_length as usize)];
+        self.acc += index.data_length as usize;
+        self.cursor += 1;
+        Some(data.to_vec())
```

Comment:
Going to repeat my comment from the other PR here because it was not answered.

Is this MessageSet going to be big? is performance an issue here? because it is allocating a new (short-lived it seems) vector on each iteration, and I think thats something that could be fixed (with https://docs.rs/streaming-iterator/ for example).

Reply:
the size will depend on the mount of data being fetched, i am reading the data in irregular sizes i dont know how the streaming iterator can help, but if you can provide me a example i will have a look at it

Reply:
Let's say the `MessageSet` contains 100 entries, if we iterate it with the current implementation, we will perform 100 allocations to copy from the chunks of data we already have. Now, if we process 100 `MessageSet`s, that is 100*100 allocations.

With the streaming iterator we would instead just borrow those slices, avoiding the need to perform allocation here.

Reply:
cc @bkontur

Reply:
Another option (assuming the `MessageSet` is not used after it gets iterated over, because this would mutate it) could be to use Vec's `drain` method https://doc.rust-lang.org/std/vec/struct.Vec.html#method.drain

Reply:
ok, @bkontur said:

> we usually read by one item (read one BlockHeader’s data for RPC request), but I think there are some list rpcs for headers, I hope that means just “few”

If that is the case I guess we are good for now (I would still like to optimize this, but not important for now, so don't worry about it).

Reply:
we only read 50 headers at time